src/v/coproc/script_database.h

```diff
@@ -0,0 +1,112 @@
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "coproc/types.h"
+
+#include <absl/container/node_hash_map.h>
+#include <absl/container/node_hash_set.h>
+
+#include <vector>
+
+namespace coproc::wasm {
+/// In memory cache of all coprocessors of all coprocessors
+///
+/// Useful when its necessary to redeploy or reference a shutdown or partitally
+/// shutdown coprocessor for its metadata
+class script_database {
+public:
+    struct script_metadata {
+        iobuf source;
+        std::vector<topic_namespace_policy> inputs;
+    };
+
+    using underlying_t = absl::node_hash_map<script_id, script_metadata>;
+    using inverted_lookup_t = absl::node_hash_map<
+      model::topic_namespace,
+      absl::node_hash_set<script_id>,
+      model::topic_namespace_hash,
+      model::topic_namespace_eq>;
+    using const_iterator = underlying_t::const_iterator;
+    using const_inv_iterator = inverted_lookup_t::const_iterator;
+
+    /// Querying
+    /// ... by scripts
+    bool exists(script_id id) { return _db.find(id) != _db.end(); }
+    const const_iterator find(script_id id) const { return _db.find(id); }
+    const const_iterator cbegin() const { return _db.cbegin(); }
+    const const_iterator cend() const { return _db.cend(); }
+
+    /// ... by topic/namespace
+    bool exists(model::topic_namespace_view tnv) const {
+        return _by_input.find(tnv) != _by_input.cend();
+    }
+    const const_inv_iterator find(model::topic_namespace_view tnv) const {
+        return _by_input.find(tnv);
+    }
+    const const_inv_iterator inv_cbegin() const { return _by_input.cbegin(); }
+    const const_inv_iterator inv_cend() const { return _by_input.cend(); }
+
+    underlying_t::size_type size() const { return _db.size(); }
+    inverted_lookup_t::size_type unique_inputs() const {
+        return _by_input.size();
+    }
+
+    /// Removal
+    bool remove_script(script_id id) {
+        auto found = _db.find(id);
+        if (found == _db.end()) {
+            return false;
+        }
+        for (const auto& e : found->second.inputs) {
+            _by_input.erase(e.tn);
+        }
+        _db.erase(found);
+        return true;
+    }
+    void clear() {
+        _db.clear();
+        _by_input.clear();
+    }
+
+    /// Addition
+    bool add_script(
+      script_id id, iobuf source, std::vector<topic_namespace_policy> inputs) {
+        script_metadata meta{.source = std::move(source), .inputs = inputs};
+        auto [_, success] = _db.emplace(id, std::move(meta));
+        if (!success) {
+            return false;
+        }
+        for (auto& input : inputs) {
+            auto found = _by_input.find(input.tn);
+            if (found == _by_input.end()) {
+                absl::node_hash_set<script_id> ns;
+                ns.emplace(id);
+                _by_input.emplace(std::move(input.tn), std::move(ns));
+            } else {
+                found->second.emplace(id);
+            }
+        }
+        return true;
+    }
```

Comment:
It's tricky to make this exception safe. Can you explain the expected usage patterns of this `script_database`?

In the past I've used `boost::multi_index_container`.

Reply:
Nice food for thought, I will think about other options. To answer the question this class has 2 maps:
1. map<id -> list<topic>>
2. map<topic, list<id>>

The purpose of this is to lookup what input topics a script is interested, or do the inverse, depending on an input topic lookup which scripts were registered for this. The second map is particularly useful for partition movement, because when a source topic is moved/re-inserted it is necessary to know which co-processors care about this event.

Reply:
Maybe @VadimPlh can help here. I'm noticing I actually never invoke `script_database::find(script_id)`. I think I included it as I thought he might want to perform lookups by `script_id`. @VadimPlh do you need to do that for data policies ever? I remember seeing a map of scripts in one of your recent PRs

Reply:
If it's a lookup that runs rarely, and the number of entries is fewer than say, 100K a linear search is often good enough, and that might inform whether you need stable iterators on the primary data structure.